pmd-core/src/main/java/net/sourceforge/pmd/lang/ast/Node.java

```diff
@@ -277,7 +250,38 @@
      * the Node AST, thereby allowing tools which can operate upon DOM to also
      * indirectly operate on the AST.
      */
-    Document getAsDocument();
+    default Document getAsDocument() {
+        try {
+            final DocumentBuilderFactory dbf = DocumentBuilderFactory.newInstance();
+            final DocumentBuilder db = dbf.newDocumentBuilder();
+            final Document document = db.newDocument();
+            appendElement(document);
+            return document;
+        } catch (final ParserConfigurationException pce) {
+            throw new RuntimeException(pce);
+        }
+    }
+
+    default void appendElement(final org.w3c.dom.Node parentNode) {
```

Comment:
we wouldn't want this to be public. Unfortunately, private default methods are only introduced in Java 9, so for the time being, you would have to have this logic embedded in `getAsDocument`

Reply:
We thought that would be the case, we will merge both methods and left a comment regarding this modularization

Reply:
After seeing them merged we prefer to have both Node implementations with `getAsDocument` over nasty code.

Reply:
Why not? Internally `appendElement` is called for each `AbstractNode` [here](https://github.com/pmd/pmd/blob/master/pmd-core/src/main/java/net/sourceforge/pmd/lang/ast/AbstractNode.java#L338)

We tried to inline `appendElement` but given the method is recursive, we would prefer to have it duplicated over nasty coupled for blocks

Reply:
You are opening up the visibility of this method from protected to public, when there are no real use cases outside of this very class… this would be a prime case for private default methods, but we are siting at Java 8 at the moment.

If we can't inline it due to recursion, we may want to move it to a helper class in the same package, and have it as a static package-private method to keep it hidden from the outside world